File: MyFirstAPP/主界面/LibraryCollectionViewController.m
//
//  LibraryCollectionViewController.m
//  MyFirstAPP
//  Created by 薛超 on 16/6/24.
//  Copyright © 2016年 薛超. All rights reserved.
#import "LibraryCollectionViewController.h"
@interface LibraryCollectionViewCell : BaseCollectionViewCell
@end
@implementation LibraryCollectionViewCell
-(void)initUI{
    [super initUI];
    self.icon.frame = CGRectMake(0, 0, APPW/5, 60);
    self.icon.layer.cornerRadius = 30;
    self.icon.clipsToBounds = YES;
    self.title.frame = CGRectMake(0, YH(self.icon), W(self.icon), 20);
    self.backgroundColor = [UIColor clearColor];
    [self addSubviews:self.icon,self.title,nil];
}
-(void)setCollectionDataWithDic:(NSDictionary *)dict{
    self.title.text = [dict valueForKey:@"name"];
    self.icon.image = [UIImage imageNamed:dict[@"picture"]];
}
@end
@interface LibraryCollectionViewController (){
    NSMutableArray *libraryBooks;
}
@end
@implementation LibraryCollectionViewController
@synthesize contentLength, dismissByBackgroundDrag, dismissByBackgroundTouch, dismissByForegroundDrag;
-(id)initWithCoder:(NSCoder *)aDecoder{
    self = [super initWithCoder:aDecoder];
    if (self) {
        self.contentLength = kScreenWidth;
        self.dismissByBackgroundTouch   = YES;
        self.dismissByBackgroundDrag    = YES;
        self.dismissByForegroundDrag    = YES;
    }return self;
}


static NSString * const reuseIdentifier = @"Cell";
- (void)viewDidLoad {
    [super viewDidLoad];
    UICollectionViewFlowLayout *layout = [[UICollectionViewFlowLayout alloc]init];
    layout.itemSize = CGSizeMake(APPW/5, 80);
    layout.minimumInteritemSpacing = 10;
    layout.scrollDirection = UICollectionViewScrollDirectionHorizontal;
    layout.minimumLineSpacing = 10;
    layout.sectionInset = UIEdgeInsetsMake(30, 10, 0, 10);
    self.collectionView = [[UICollectionView alloc]initWithFrame:CGRectMake(0, 0, APPW, APPH-110) collectionViewLayout:layout];
    self.collectionView.backgroundColor = whitecolor;
    UIImageView *backView = [[UIImageView alloc]initWithFrame:CGRectMake(0, 100, kScreenWidth, kScreenHeight-100)];
    backView.image = [UIImage imageNamed:@""];
    self.collectionView.backgroundView = backView;
    ElasticTransition *ET = (ElasticTransition*)self.transitioningDelegate;
    NSLog(@"\ntransition.edge = %@\ntransition.transformType = %@\ntransition.sticky = %@\ntransition.showShadow = %@", [HelperFunctions typeToStringOfEdge:ET.edge], [ET transformTypeToString], ET.sticky ? @"YES" : @"NO", ET.showShadow ? @"YES" : @"NO");
    self.collectionView.delegate = self;
    self.collectionView.dataSource = self;
    libraryBooks = [NSMutableArray arrayWithObjects:@{@"name":@"测试数据",@"picture":@"weixinIcon"},@{@"name":@"测试数据",@"picture":@"weixinIcon"},@{@"name":@"测试数据",@"picture":@"weixinIcon"},@{@"name":@"测试数据",@"picture":@"weixinIcon"},@{@"name":@"测试数据",@"picture":@"weixinIcon"},@{@"name":@"测试数据",@"picture":@"weixinIcon"},@{@"name":@"测试数据",@"picture":@"weixinIcon"},@{@"name":@"测试数据",@"picture":@"weixinIcon"},@{@"name":@"测试数据",@"picture":@"weixinIcon"},@{@"name":@"测试数据",@"picture":@"weixinIcon"},@{@"name":@"测试数据",@"picture":@"weixinIcon"},@{@"name":@"测试数据",@"picture":@"weixinIcon"},@{@"name":@"测试数据",@"picture":@"weixinIcon"},@{@"name":@"测试数据",@"picture":@"weixinIcon"}, nil];
    [self.collectionView registerClass:[LibraryCollectionViewCell class] forCellWithReuseIdentifier:reuseIdentifier];
}
#pragma mark <UICollectionViewDataSource>

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return 1;
}
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return libraryBooks.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    LibraryCollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:reuseIdentifier forIndexPath:indexPath];
    if(!cell){
        cell = [[LibraryCollectionViewCell alloc]initWithFrame:CGRectMake(0, 0, APPW/5, 80)];
    }
    [cell setCollectionDataWithDic:libraryBooks[indexPath.item]];
    return cell;
}

#pragma mark <UICollectionViewDelegate>
- (BOOL)collectionView:(UICollectionView *)collectionView shouldHighlightItemAtIndexPath:(NSIndexPath *)indexPath {
	return YES;
}

- (BOOL)collectionView:(UICollectionView *)collectionView shouldSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    return YES;
}
- (BOOL)collectionView:(UICollectionView *)collectionView shouldShowMenuForItemAtIndexPath:(NSIndexPath *)indexPath {
	return NO;
}

- (BOOL)collectionView:(UICollectionView *)collectionView canPerformAction:(SEL)action forItemAtIndexPath:(NSIndexPath *)indexPath withSender:(id)sender {
	return NO;
}

- (void)collectionView:(UICollectionView *)collectionView performAction:(SEL)action forItemAtIndexPath:(NSIndexPath *)indexPath withSender:(id)sender {
	
}
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath{
    return CGSizeMake(240, 100);
}
@end

File: Freedom/主界面/LibraryCollectionViewController.swift
//
//  LibraryCollectionViewController.swift
//  Freedom
//
//  Created by Super on 6/15/18.
//  Copyright © 2018 薛超. All rights reserved.
//
import UIKit
import ElasticTransitionObjC
import BaseFile
import XExtension
import SVProgressHUD
class LibraryCollectionViewCell: BaseCollectionViewCell {
    override func initUI() {
        super.initUI()
        icon.frame = CGRect(x: 0, y: 0, width: APPW / 5, height: 80)
        icon.layer.cornerRadius = 40
        icon.clipsToBounds = true
        title.frame = CGRect(x: 0, y: YH(icon), width: W(icon), height: 20)
        title.textAlignment = .center
        backgroundColor = UIColor.clear
        addSubviews([icon, title])
    }
}
class LibraryCollectionViewController: UICollectionViewController,ElasticMenuTransitionDelegate {
    var contentLength:CGFloat = APPW
    var dismissByBackgroundTouch = true
    var dismissByBackgroundDrag = true
    var dismissByForegroundDrag = true
    override func viewDidLoad() {
        super.viewDidLoad()
        let layout = UICollectionViewFlowLayout()
        layout.itemSize = CGSize(width: APPW / 5, height: 100)
        layout.minimumInteritemSpacing = 10
        layout.scrollDirection = .vertical
        layout.minimumLineSpacing = 30
        layout.sectionInset = UIEdgeInsetsMake(30, 10, 0, 10)
        collectionView = UICollectionView(frame: CGRect(x: 0, y: 0, width: APPW, height: APPH - 110), collectionViewLayout: layout)
        collectionView?.backgroundColor = UIColor.white
        let backView = UIImageView(frame: CGRect(x: 0, y: 100, width: APPW, height: APPH - 100))
        backView.image = UIImage(named: "")
        collectionView?.backgroundView = backView
        let ET = transitioningDelegate as? ElasticTransition
        print("\ntransition.edge = \(HelperFunctions.type(toStringOf: (ET?.edge)!))\ntransition.transformType = \(String(describing: ET?.transformTypeToString()))\ntransition.sticky = \(String(describing:  ET?.sticky))\ntransition.showShadow = \(String(describing: ET?.showShadow))")
        collectionView?.register(LibraryCollectionViewCell.self, forCellWithReuseIdentifier: LibraryCollectionViewCell.getColloctionCellIdentifier())
        collectionView?.delegate = self
        collectionView?.dataSource = self
    }
    override func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 1
    }
    override func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return FreedomItems.count
    }
    override func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        var cell = collectionView.dequeueReusableCell(withReuseIdentifier: LibraryCollectionViewCell.getColloctionCellIdentifier(), for: indexPath) as? LibraryCollectionViewCell
        if cell == nil {
            cell = LibraryCollectionViewCell(frame: CGRect(x: 0, y: 0, width: APPW / 5, height: 100))
        }
        let dict = FreedomItems[indexPath.row]
        cell?.title.text = dict["title"]
        cell?.icon.image = UIImage(named: dict["icon"]!)
        if let aCell = cell {
            return aCell
        }
        return UICollectionViewCell()
    }
    override func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        let controlName = FreedomItems[indexPath.row]["control"]!
        UIApplication.shared.isStatusBarHidden = false
        if !(controlName == "Sina") {
            let StoryBoard = UIStoryboard(name: controlName, bundle: nil)
            show(StoryBoard.instantiateViewController(withIdentifier: "\(controlName)TabBarController"), sender: self)
            return
        }
        let s = "\(controlName)TabBarController"
        print(s)
        let con: UIViewController = UIViewController()//NSClassFromString(s)
        let animation = CATransition()
        animation.duration = 1
        animation.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)
        view.window?.layer.add(animation, forKey: nil)
        present(con, animated: false) {
            SVProgressHUD.showSuccess(withStatus: FreedomItems[indexPath.row]["title"])
        }
    }
}
